src/components/fields/RectanglePositioner.js

```diff
@@ -0,0 +1,194 @@
+    this.attr = this.props.cartesian
+      ? {
+          x: ['xaxis.domain[0]', 'xaxis.domain[1]'],
+          y: ['yaxis.domain[0]', 'yaxis.domain[1]'],
+        }
+      : {x: ['domain.x[0]', 'domain.x[1]'], y: ['domain.y[0]', 'domain.y[1]']};
+    this.state = {snap: true};
+  }
+
+  sendUpdate({x, y, width, height, fieldWidthPx, fieldHeightPx}) {
+    const x0 = x / fieldWidthPx;
+    const x1 = (width + x) / fieldWidthPx;
+    const y0 = (fieldHeightPx - (height + y)) / fieldHeightPx;
+    const y1 = (fieldHeightPx - y) / fieldHeightPx;
+
+    const snap = this.state.snap
+      ? v => Math.round(v * gridRes) / gridRes
+      : v => v;
+
+    const payload = {};
+
+    if (x0 >= 0 && x1 <= 1) {
+      payload[this.attr.x[0]] = snap(x0);
+      payload[this.attr.x[1]] = snap(x1);
+    }
+
+    if (y0 >= 0 && y1 <= 1) {
+      payload[this.attr.y[0]] = snap(y0);
+      payload[this.attr.y[1]] = snap(y1);
+    }
+
+    this.context.updateContainer(payload);
+  }
+
+  render() {
+    const {attr, cartesian} = this.props;
+    const {
+      localize: _,
+      fullContainer,
+      fullLayout: {width: plotWidthPx, height: plotHeightPx},
+    } = this.context;
+    const x = cartesian ? fullContainer.xaxis.domain : fullContainer.domain.x;
+    const y = cartesian ? fullContainer.yaxis.domain : fullContainer.domain.y;
+    const aspectRatio = plotHeightPx / plotWidthPx;
+    const fieldWidthPx = Math.min(maxWidth, maxWidth / aspectRatio);
+    const fieldHeightPx = Math.min(maxWidth, maxWidth * aspectRatio);
+
+    const width = fieldWidthPx * (x[1] - x[0]);
+    const height = fieldHeightPx * (y[1] - y[0]);
+    const left = fieldWidthPx * x[0];
+    const top = fieldHeightPx * (1 - y[1]);
+
+    let zoomable = '';
+    if (
+      !fullContainer.xaxis ||
+      !fullContainer.yaxis ||
+      (!fullContainer.xaxis.overlaying && !fullContainer.yaxis.overlaying)
+    ) {
+      zoomable = 'n, w, s, e, nw, ne, se, sw';
+    } else if (!fullContainer.xaxis.overlaying) {
+      zoomable = 'e, w';
+    } else if (!fullContainer.yaxis.overlaying) {
+      zoomable = 'n, s';
+    }
+
+    return (
+      <Field {...this.props} attr={attr}>
+        <Field label={_('Snap to Grid')}>
+          <RadioBlocks
+            alignment="center"
+            onOptionChange={snap => this.setState({snap: snap})}
+            activeOption={this.state.snap}
+            options={[
+              {label: _('On'), value: true},
+              {label: _('Off'), value: false},
+            ]}
+          />
+        </Field>
+        <div
+          className="rect-container"
+          style={{
+            width: fieldWidthPx + 1,
+            height: fieldHeightPx + 1,
+          }}
+        >
+          {Array(gridRes * gridRes)
+            .fill(0)
+            .map((v, i) => (
+              <div
+                key={i}
+                className="rect-grid"
+                style={{
+                  width: fieldWidthPx / gridRes - 1,
```

Comment:
could we move this to a style sheet?

Reply:
not really, if you see the math we're doing here, it doesn't really map well onto a stylesheet

Reply:
I suppose we could create classes for the various grid-box-types but there are 9 of them (4 corners, 4 faces, inside) which would be messier than just building them up from stratch like this IMO. Although with faaaancy CSS3 selectors you might be able to do it ;)